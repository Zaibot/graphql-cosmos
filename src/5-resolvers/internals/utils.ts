import { FeedResponse } from '@azure/cosmos'
import { GraphQLCosmosPageOutput } from '../../4-resolver-builder/3-typedefs-transformer'
import { Lazy } from '../../typescript'
import { SourceDescriptor } from '../x-descriptors'

export const emptyPageResponse: GraphQLCosmosPageOutput = {
  cursor: null,
  nextCursor: null,
  page: [],
  total: 0,
}

export function graphqlCosmosPageResponse(
  typename: string,
  database: string,
  container: string,
  cursor: string | null,
  query: Lazy<Promise<FeedResponse<{ id: string }>>>,
  count: Lazy<Promise<FeedResponse<number>>>
): GraphQLCosmosPageOutput {
  return {
    get cursor() {
      return Promise.resolve(cursor)
    },
    get nextCursor() {
      return query().then((x) => x.continuationToken)
    },
    get page() {
      return query().then((x) => x.resources.map(wrapSingleSourceDescriptor(typename, database, container)))
    },
    get total() {
      return count().then((x) => x.resources[0])
    },
  }
}

export function wrapSingleSourceDescriptor(typename: string, database: string, container: string) {
  // TODO, use refSingle of datasource
  return (item: { id: string }) =>
    SourceDescriptor.withDescriptor(item, {
      kind: `Single`,
      database,
      container,
      typename,
      id: item.id,
    })
}
