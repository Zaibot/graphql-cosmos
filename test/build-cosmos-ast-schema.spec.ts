import gql from 'graphql-tag'
import { printSchemaWithDirectives } from '@graphql-tools/utils'
import { createUnitTestContext } from './utils'

describe(`Build Cosmos AST Schema`, () => {
  const dummyTypeDefs = gql`
    type Query {
      dummies: [Dummy!]! @cosmos
    }
    type Dummy @cosmos(database: "Test", container: "Dummies") {
      id: ID! @where(op: "eq")
      status: Status! @where(op: "eq neq")
      related: [Related!]! @cosmos(ours: "relatedIds", pagination: "on")
    }

    type Related @cosmos(database: "Test", container: "Relations") {
      id: ID! @where(op: "eq")
      name: String! @sort(ours: "test")
    }

    enum Status {
      OPEN
      CLOSE
    }
  `

  const responses = {}

  const uc = createUnitTestContext(dummyTypeDefs, responses)

  it(`expects schema to remain the same`, () => {
    const output = printSchemaWithDirectives(uc.schema)
    expect(output).toMatchSnapshot()
  })

  it(`expects meta schema to remain the same`, () => {
    const output = uc.metaSchema
    expect(output).toMatchSnapshot()
  })
})
