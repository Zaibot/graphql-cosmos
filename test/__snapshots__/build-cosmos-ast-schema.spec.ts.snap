// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build Cosmos AST Schema expects meta schema to remain the same 1`] = `
Object {
  "types": Array [
    Object {
      "container": "Dummies",
      "cosmos": true,
      "database": "Test",
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
          ],
          "whereOurs": null,
        },
        Object {
          "container": "Relations",
          "cosmos": true,
          "database": "Test",
          "fieldname": "related",
          "ours": "relatedIds",
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Related",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "status",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Status",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
            "neq",
          ],
          "whereOurs": null,
        },
      ],
      "filterable": true,
      "sortable": false,
      "typename": "Dummy",
    },
    Object {
      "container": null,
      "cosmos": false,
      "database": null,
      "external": false,
      "fields": Array [
        Object {
          "container": "Dummies",
          "cosmos": true,
          "database": "Test",
          "fieldname": "dummies",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Dummy",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Query",
    },
    Object {
      "container": "Relations",
      "cosmos": true,
      "database": "Test",
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
          ],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "name",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "String",
          "sortOurs": "test",
          "sortable": true,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": true,
      "sortable": true,
      "typename": "Related",
    },
  ],
}
`;

exports[`Build Cosmos AST Schema expects schema to remain the same 1`] = `
"schema {
  query: Query
}

type Query {
  dummies(where: DummyWhere, cursor: String, limit: Int): DummyPage!
}

type Dummy {
  id: ID!
  status: Status!
  related(where: RelatedWhere, sort: RelatedSort, cursor: String, limit: Int): RelatedPage!
}

type Related {
  id: ID!
  name: String!
}

enum Status {
  OPEN
  CLOSE
}

type RelatedPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Related]!
}

type DummyPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Dummy]!
}

input DummyWhere {
  and: [DummyWhere!]
  or: [DummyWhere!]
  id_eq: ID
  status_eq: Status
  status_neq: Status
}

input RelatedWhere {
  and: [RelatedWhere!]
  or: [RelatedWhere!]
  id_eq: ID
}

input RelatedSort {
  name_ASC: Int
  name_DESC: Int
}
"
`;
